### ISSUE TITLE
References to registers that are not restored during "full-dynamic trace" cause unintended problems.

### ISSUE THREAD

**Issue by ParkHanbum** [namhyung/uftrace#731]
this is related with https://github.com/namhyung/uftrace/issues/281#issuecomment-477950844. The problem issued by @honggyukim 

This is probably related to #372 as well.

As we have already discussed, there is an unintended problem with "full-dynamic trace".
I want to talking about it. 

# Symptom

as @honggyukim said, Tracing uftrace itself with full-dynamic will cause a segment fault. 

```
$ ./uftrace record -L. -P. -v -la --no-libcall ./uftrace replay -d abc --no-pager
```

this command will make following results.
```
$ ./uftrace record -L. -P. -v -la --no-libcall ./uftrace replay -d abc --no-pager
uftrace: skipping perf event due to error: Permission denied
uftrace: using ./libmcount/libmcount.so library for tracing
uftrace: please check /proc/sys/kernel/perf_event_paranoid
uftrace: creating 1 thread(s) for recording
mcount: initializing mcount library
dwarf: prepare debug info
dynamic: dynamic patch type: 0 (none)
uftrace: some functions cannot be patched dynamically (581/751)
dynamic: dynamic update stats:
dynamic:    total:      751
dynamic:  patched:      581 (77.36%)
dynamic:   failed:      161 (21.44%)
dynamic:  skipped:        9 (1.20%)
dynamic: no match:        0
mcount: mcount setup done
mcount: new session started: 4a5c6a90d62ffbf2: uftrace
Segmentation fault: address not mapped (addr: 0x1)
Backtrace from uftrace:
=====================================
[8] (rb_insert_color[55fa20f0aa45] <= add_trigger_entry.isra.5[55fa20f04330])
[7] (add_trigger_entry.isra.5[55fa20f040d5] <= setup_trigger[55fa20f04d67])
[6] (setup_trigger[55fa20f04825] <= mcount_return[7f1ec5ce081f])
[5] (uftrace_setup_trigger[55fa20f04e45] <= build_fixup_filter[55fa20f05657])
[4] (build_fixup_filter[55fa20f055d5] <= walk_sessions[55fa20ef46e9])
[3] (walk_sessions[55fa20ef46b5] <= fstack_setup_filters[55fa20f060f0])
[2] (fstack_setup_filters[55fa20f05b75] <= command_replay[55fa20ee5899])
[1] (command_replay[55fa20ee5825] <= main[55fa20edbd9a])
[0] (main[55fa20edb7e5] <= __libc_start_main[7f1ec4675b97])
child terminated by signal: 11: Segmentation fault
uftrace: reading uftrace.data/task.txt file
uftrace: flushing /uftrace-4a5c6a90d62ffbf2-31158-001
```

it is little-diff between @honggyukim 's comment. so, it may be a different problem. 
Anyway, I have traced the cause of problem. and found some weird instruction that created by compiler. 

# Where the problem occurs
in function __rb_rotate_left : 

following disassembled instructions did not occured any problem.
```asm
   3d729:       e8 72 fe ff ff          callq  3d5a0 <__rb_rotate_left>
   3d72e:       48 8b 04 24             mov    (%rsp),%rax
   3d732:       48 8b 28                mov    (%rax),%rbp
   3d735:       48 83 e5 fc             and    $0xfffffffffffffffc,%rbp
   3d739:       48 85 ed                test   %rbp,%rbp
   3d73c:       48 89 ee                mov    %rbp,%rsi
   3d73f:       75 9c                   jne    3d6dd <rb_insert_color+0x1d>
   3d741:       48 8b 44 24 08          mov    0x8(%rsp),%rax
   3d746:       48 8b 00                mov    (%rax),%rax
   3d749:       48 83 08 01             orq    $0x1,(%rax)
```
compare with following disassembled instruction that cause the problem.
```asm
   39a96:       e8 85 fe ff ff          callq  39920 <__rb_rotate_left>
   39a9b:       4c 8b 0b                mov    (%rbx),%r9
   39a9e:       49 83 e1 fc             and    $0xfffffffffffffffc,%r9
   39aa2:       4d 85 c9                test   %r9,%r9
   39aa5:       4c 89 c9                mov    %r9,%rcx
   39aa8:       75 ac                   jne    39a56 <rb_insert_color+0x16>
   39aaa:       49 8b 03                mov    (%r11),%rax
   39aad:       48 83 08 01             orq    $0x1,(%rax)
```
did you found where the problem is?

here is important part : 
```asm
instructions that did not cause problem:
   3d741:       48 8b 44 24 08          mov    0x8(%rsp),%rax
   3d746:       48 8b 00                mov    (%rax),%rax
   3d749:       48 83 08 01             orq    $0x1,(%rax)
instructions that cause problem:
   39aaa:       49 8b 03                mov    (%r11),%rax
```

The instructions causing the problem refer directly to registers that were not restored after the function call. On the other hand, instructions that did not causing the problems were load the values ​​stored in the stack and refer to them.

this is same type issue of #372 .

# the cause of a problem
As I already said, this is related to the "ipa-ra" optimization.
This is the definition of the "ipa-ra" option.

```
-fipa-ra

Use caller save registers for allocation if those registers are not used by any called 
function. In that case it is not necessary to save and restore them around calls. This 
is only possible if called functions are part of same compilation unit as current 
function and they are compiled before it.

Enabled at levels -O2, -O3, -Os, however the option is disabled if generated code
will be instrumented for profiling (-p, or -pg) or if callee’s register usage cannot be
known exactly (this happens on targets that do not expose prologues and 
epilogues in RTL).

```
In summary, it seems to improve performance by allocating the same variable that is duplicated during a function call to a register.

# Diagnosis
1. The same variable that is duplicated during the function call is assigned to the register.
2. Using full-dynamic trace, it will be call mcount-entry, mcount_find_code, and mcount_return, which are not calcuated in compile-time.
3. mcount-entry, mcount_find_code, mcount_return If a register value assigned to a variable changes during a call, an unintended change in code flow occurs.

ps. If you use an option such as '-pg', the option is disabled.

# Confirmation
There are three ways to prove that the diagnosis for this problem is correct.
### 1. disable 'ipa-ra'
```
$ git status
On branch dynamic-trace-v6
Your branch is up to date with 'origin/review/dynamic-trace-v6'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   Makefile

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        abc/
        cmd.gdb
        t-abc

no changes added to commit (use "git add" and/or "git commit -a")
```
```diff
$ git diff
diff --git a/Makefile b/Makefile
index fa00e06..85e79a2 100644
--- a/Makefile
+++ b/Makefile
@@ -82,7 +82,7 @@ TEST_LDFLAGS       = $(COMMON_LDFLAGS) -L$(objdir)/libtraceevent -ltraceevent
 ifeq ($(DEBUG), 1)
   COMMON_CFLAGS += -O0 -g
 else
-  COMMON_CFLAGS += -O2 -g
+  COMMON_CFLAGS += -O2 -g -fno-ipa-ra
 endif

 ifeq ($(TRACE), 1)

```
```
$ ./uftrace record -L. -P. -la --no-libcall ./uftrace replay -d abc --no-pager
uftrace: some functions cannot be patched dynamically (581/751)
# DURATION     TID     FUNCTION
   0.564 us [  3923] | __monstartup();
   0.325 us [  3923] | __cxa_atexit();
            [  3923] | main() {
            [  3923] |   a() {
            [  3923] |     b() {
            [  3923] |       c() {
   0.515 us [  3923] |         getpid();
   0.784 us [  3923] |       } /* c */
   0.996 us [  3923] |     } /* b */
   1.149 us [  3923] |   } /* a */
   1.284 us [  3923] | } /* main */
```

### 2. use '-ffixed-<reg_name>'
```diff
$ git diff
diff --git a/Makefile b/Makefile
index fa00e06..b842d97 100644
--- a/Makefile
+++ b/Makefile
@@ -61,6 +61,13 @@ endif

 COMMON_CFLAGS += -W -Wall -Wno-unused-parameter -Wno-missing-field-initializers

+ifeq ($(DEBUG), 1)
+  COMMON_CFLAGS += -O0 -g
+else
+  COMMON_CFLAGS += -O2 -g -ffixed-rbx -ffixed-r8 -ffixed-r9 -ffixed-r10 -ffixed-r11 -ffixed-r12 -ffixed-r13 -ffixed-r14 -ffixed-r15
+endif
+
+
 #
 # Note that the plain CFLAGS and LDFLAGS can be changed
 # by config/Makefile later but *_*FLAGS can not.
@@ -79,11 +86,6 @@ SYMBOLS_LDFLAGS    = $(COMMON_LDFLAGS) $(LDFLAGS_$@) $(LDFLAGS_symbols)
 LIB_LDFLAGS        = $(COMMON_LDFLAGS) $(LDFLAGS_$@) $(LDFLAGS_lib) -Wl,--no-undefined
 TEST_LDFLAGS       = $(COMMON_LDFLAGS) -L$(objdir)/libtraceevent -ltraceevent

-ifeq ($(DEBUG), 1)
-  COMMON_CFLAGS += -O0 -g
-else
-  COMMON_CFLAGS += -O2 -g
-endif

 ifeq ($(TRACE), 1)
   UFTRACE_CFLAGS    += -pg -fno-omit-frame-pointer
```
```
$ ./uftrace record -L. -P. -la --no-libcall ./uftrace replay -d abc --no-pager
uftrace: some functions cannot be patched dynamically (572/751)
# DURATION     TID     FUNCTION
   0.564 us [  3923] | __monstartup();
   0.325 us [  3923] | __cxa_atexit();
            [  3923] | main() {
            [  3923] |   a() {
            [  3923] |     b() {
            [  3923] |       c() {
   0.515 us [  3923] |         getpid();
   0.784 us [  3923] |       } /* c */
   0.996 us [  3923] |     } /* b */
   1.149 us [  3923] |   } /* a */
   1.284 us [  3923] | } /* main */
```

### 3. store/restore all of registers in mcount
```
$ git status
On branch dynamic-trace-v6
Your branch is up to date with 'origin/review/dynamic-trace-v6'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   arch/x86_64/dynamic.S
        modified:   arch/x86_64/mcount.S

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        abc.old/
        abc/
        cmd.gdb
        q
        t-abc

no changes added to commit (use "git add" and/or "git commit -a")
```

```diff
$ git diff
diff --git a/arch/x86_64/dynamic.S b/arch/x86_64/dynamic.S
index ec19260..c9dff6d 100644
--- a/arch/x86_64/dynamic.S
+++ b/arch/x86_64/dynamic.S
@@ -42,9 +42,16 @@

 GLOBAL(__dentry__)
        .cfi_startproc
-       sub $48, %rsp
+       sub $104, %rsp
        .cfi_adjust_cfa_offset 48

+       movq %r15, 96(%rsp)
+       movq %r14, 88(%rsp)
+       movq %r13, 80(%rsp)
+       movq %r12, 72(%rsp)
+       movq %r11, 64(%rsp)
+       movq %r10, 56(%rsp)
+       movq %rbx, 48(%rsp)
        movq %rdi, 40(%rsp)
        movq %rsi, 32(%rsp)
        movq %rdx, 24(%rsp)
@@ -53,10 +60,10 @@ GLOBAL(__dentry__)
        movq %r9,   0(%rsp)

        /* child addr */
-       movq 48(%rsp), %rsi
+       movq 104(%rsp), %rsi

        /* parent location */
-       lea 56(%rsp), %rdi
+       lea 112(%rsp), %rdi

        /* mcount_args */
        movq %rsp, %rdx
@@ -75,7 +82,7 @@ GLOBAL(__dentry__)
        movq 8(%rsp), %rdx

        /* child addr */
-       movq 48(%rdx), %rdi
+       movq 104(%rdx), %rdi

        /* find location that has the original code */
        call mcount_find_code
@@ -84,7 +91,7 @@ GLOBAL(__dentry__)
        movq 8(%rsp), %rdx

        /* overwrite return address */
-       movq %rax, 48(%rdx)
+       movq %rax, 104(%rdx)

        pop  %rax

@@ -97,8 +104,15 @@ GLOBAL(__dentry__)
        movq 24(%rsp), %rdx
        movq 32(%rsp), %rsi
        movq 40(%rsp), %rdi
-
-       add $48, %rsp
+       movq 48(%rsp), %rbx
+       movq 56(%rsp), %r10
+       movq 64(%rsp), %r11
+       movq 72(%rsp), %r12
+       movq 80(%rsp), %r13
+       movq 88(%rsp), %r14
+       movq 96(%rsp), %r15
+
+       add $104, %rsp
        .cfi_adjust_cfa_offset -48

        retq

diff --git a/arch/x86_64/mcount.S b/arch/x86_64/mcount.S
index 283b9c2..d996071 100644
--- a/arch/x86_64/mcount.S
+++ b/arch/x86_64/mcount.S
@@ -84,9 +84,21 @@ END(mcount)

 ENTRY(mcount_return)
        .cfi_startproc
-       sub $48, %rsp
+       sub $144, %rsp
        .cfi_def_cfa_offset 48

+       movq   %r15,  128(%rsp)
+       movq   %r14,  120(%rsp)
+       movq   %r13,  112(%rsp)
+       movq   %r12,  104(%rsp)
+       movq   %r11,  96(%rsp)
+       movq   %r10,  88(%rsp)
+       movq   %r9,   80(%rsp)
+       movq   %r8,   72(%rsp)
+       movq   %rdi,  64(%rsp)
+       movq   %rsi,  56(%rsp)
+       movq   %rcx,  48(%rsp)
+       movq   %rbx,  40(%rsp)
        movdqu %xmm0, 16(%rsp)
        movq   %rdx,   8(%rsp)
        movq   %rax,   0(%rsp)
@@ -96,13 +108,25 @@ ENTRY(mcount_return)

        /* returns original parent address */
        call mcount_exit
-       movq %rax, 40(%rsp)
+       movq %rax, 136(%rsp)

        movq    0(%rsp), %rax
        movq    8(%rsp), %rdx
        movdqu 16(%rsp), %xmm0
-
-       add $40, %rsp
+       movq   40(%rsp), %rbx
+       movq   48(%rsp), %rcx
+       movq   56(%rsp), %rsi
+       movq   64(%rsp), %rdi
+       movq   72(%rsp), %r8
+       movq   80(%rsp), %r9
+       movq   88(%rsp), %r10
+       movq   96(%rsp), %r11
+       movq   104(%rsp), %r12
+       movq   112(%rsp), %r13
+       movq   120(%rsp), %r14
+       movq   128(%rsp), %r15
+
+       add $136, %rsp
        .cfi_def_cfa_offset 8
        retq
        .cfi_endproc

```
```
$ ./uftrace record -L. -P. -la --no-libcall ./uftrace replay -d abc --no-pager
uftrace: some functions cannot be patched dynamically (581/751)
# DURATION     TID     FUNCTION
   0.500 us [ 78669] | __monstartup();
   0.182 us [ 78669] | __cxa_atexit();
            [ 78669] | main() {
            [ 78669] |   a() {
            [ 78669] |     b() {
            [ 78669] |       c() {
   0.356 us [ 78669] |         getpid();
   0.651 us [ 78669] |       } /* c */
   0.826 us [ 78669] |     } /* b */
   0.977 us [ 78669] |   } /* a */
   1.117 us [ 78669] | } /* main */
```

# Conclusion

As we've seen through this issue, the compiler has optimizations that assign variables to registers and share them among functions. When "full-dynamic" is used, if a register that is not restored is assigned as a variable, the effect is the same as changing the value of a variable used in the target program unintentionally.

Storing and restoring all the registers is the fastest and easiest way to solve this problem. But it sacrifices performance. If there is not need to call an external library function internally when call mcount-xxx function, there is another method that compile uftrace with "-ffixed- <reg_name>" option.

**Comment by honggyukim** [comment]
@ParkHanbum Thanks for very useful explanation! So the point is that it works fine if we save all the registers even if `-fipa-ra` is used.

If we don't want to sacrifice performance by restoring all the registers, why don't we split `libmcount.so` and `libmcount-dynamic.so`? We can restore all the registers only in `libmcount-dynamic.so`, but keep `libmcount.so` as is.

**Comment by honggyukim** [comment]
This is really cool! I've tested it again after applying the change "3. store/restore all of registers in mcount". I found that it mostly works fine and here is the well tested programs.
1. node
2. clang
3. tmux
4. uftrace (itself)

The only thing that doesn't look good is `cgdb` program. As a first step, it works fine with the below execution sequence.
```
$ uftrace record -P. `which cgdb` -ex="q" hello
```
1. launch cgdb
2. quit

However, it gets stuck with the below sequence.
```
$ uftrace record -P. `which cgdb` -ex="b main" -ex="r" -ex="q" hello
```
1. launch cgdb
2. set breakpoint to "main"
3. run
4. quit after hitting the breakpoint

After running the above test, uftrace looks running but doesn't finish.
```
$ uftrace record -P. `which cgdb` -ex="b main" -ex="r" -ex="q" hello
uftrace: some functions cannot be patched dynamically (391/785)
(... running forever ...)
```
I found that it keeps printing the below debug message while waiting.
```
    ...
uftrace: waiting for FORK2
uftrace: waiting for FORK2
uftrace: waiting for FORK2
uftrace: waiting for FORK2
uftrace: waiting for FORK2
uftrace: waiting for FORK2
uftrace: waiting for FORK2
uftrace: waiting for FORK2
    ...
```

**Comment by honggyukim** [comment]
Hmm.. I tested it again for `chrome`, but it doesn't work properly as follows:
```
$ uftrace record -P. -t 1ms ./chrome --no-sandbox --headless https://www.google.com
uftrace: some functions cannot be patched dynamically (285354/326211)
uftrace: some functions cannot be patched dynamically (285354/326211)
Received signal 4 ILL_ILLOPN 55ef4a8ae164
#0 0x7f2d98c20e47 backtrace
#1 0x55ef4cb9e14f (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca214e)
#2 0x55ef4cb9dcb1 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca1cb0)
#3 0x7f2d98806390 <unknown>
#4 0x55ef4a8ae164 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x29b2163)
#5 0x55ef4d438bde (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x553cbdd)
#6 0x55ef4d43840d (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x553c40c)
#7 0x55ef4d4667db (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x556a7da)
#8 0x55ef4d4673b9 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x556b3b8)
#9 0x55ef4d42fc01 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5533c00)
#10 0x55ef4d42c936 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5530935)
#11 0x55ef4d42ca81 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5530a80)
#12 0x55ef4d4c0753 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55c4752)
#13 0x55ef4d4bc791 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55c0790)
#14 0x55ef4d4b40fb (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55b80fa)
#15 0x55ef4d4b1804 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55b5803)
#16 0x55ef4d21aff7 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x531eff6)
#17 0x55ef4d219d3c (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x531dd3b)
#18 0x55ef4d21c1c7 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x53201c6)
#19 0x55ef4d23c701 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5340700)
#20 0x55ef4d23cb9c (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5340b9b)
#21 0x55ef4d24f732 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5353731)
#22 0x55ef4d24fcb1 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5353cb0)
#23 0x55ef4d24f8b2 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x53538b1)
#24 0x55ef4a8d1527 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x29d5526)
#25 0x55ef4b2184fe (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x331c4fd)
#26 0x55ef4cb8cf5a (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c90f59)
#27 0x55ef4caf17c3 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4bf57c2)
#28 0x55ef4cb0d91f (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c1191e)
#29 0x55ef4cb0dfe2 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c11fe1)
#30 0x55ef4cbb00e9 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4cb40e8)
#31 0x55ef4cb2e539 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c32538)
#32 0x55ef4b142c34 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x3246c33)
#33 0x55ef4cb721e2 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c761e1)
#34 0x55ef4cbb346e (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4cb746d)
#35 0x7f2d987fc6ba start_thread
#36 0x7f2d9277341d clone
  r8: 00000ccfd7d44ee8  r9: 0000000000000020 r10: 0000000000000000 r11: 0000000000080100
 r12: 0000000000000020 r13: 00000ccfd7d44ee8 r14: 00000cd0187bd3e8 r15: 00000ccfd7d44eb8
  di: 00000cd01871a488  si: 00000cd01871ad88  bp: 00007f2c9da28a50  bx: 00000cd0187bd388
  dx: 00000cd018535fc8  ax: 00007f2c9da28a18  cx: 00000cd01871aea8  sp: 00007f2c9da28a10
  ip: 000055ef4a8ae164 efl: 0000000000010246 cgf: 0000000000000033 erf: 0000000000000000
 trp: 0000000000000006 msk: 0000000000000000 cr2: 0000000000000000
[end of stack trace]
Calling _exit(1). Core file will not be generated.
uftrace: some functions cannot be patched dynamically (285354/326211)
```
But it recorded fine until it abnormally finished.
```
$ uftrace replay -t 100ms -f +task
# DURATION     TID         TASK NAME   FUNCTION
            [141297]          chrome | ChromeMain() {
            [141297]          chrome |   headless::HeadlessShellMain() {
            [141297]          chrome |     headless::HeadlessShellMain() {
            [141297]          chrome |       headless::HeadlessBrowserMain() {
            [141297]          chrome |         headless::_GLOBAL__N_1::RunContentMain() {
            [141297]          chrome |           content::ContentMain() {
            [141297]          chrome |             service_manager::Main() {
            [141297]          chrome |               /* linux:task-name (name=chrome) */
            [141297]          chrome |               content::ContentServiceManagerMainDelegate::Initialize() {
            [141297]          chrome |                 content::ContentMainRunnerImpl::Initialize() {
            [141418]          chrome | base::_GLOBAL__N_1::ThreadFunc() {
            [141418]          chrome |   base::SimpleThread::ThreadMain() {
            [141418] sandbox_ipc_thr |     /* linux:task-name (name=sandbox_ipc_thr) */
            [141418] sandbox_ipc_thr |     base::DelegateSimpleThread::Run() {
            [141418] sandbox_ipc_thr |       content::SandboxIPCHandler::Run() {
            [141418] sandbox_ipc_thr |         poll() {
            [141418] sandbox_ipc_thr |           /* linux:sched-out */
            [141297]          chrome |                   service_manager::ZygoteCommunication::GetSandboxStatus() {
            [141297]          chrome |                     read() {
            [141297]          chrome |                       /* linux:sched-out */
            [141419]          chrome | } /* fork */
            [141419]          chrome | execvp() {
            [141419]          chrome | ChromeMain() {
            [141419]          chrome |   headless::HeadlessShellMain() {
            [141419]          chrome |     headless::HeadlessShellMain() {
            [141419]          chrome |       headless::RunChildProcessIfNeeded() {
            [141419]          chrome |         headless::_GLOBAL__N_1::RunContentMain() {
            [141419]          chrome |           content::ContentMain() {
            [141419]          chrome |             service_manager::Main() {
            [141419]          chrome |               /* linux:task-name (name=chrome) */
            [141419]          chrome |               content::ContentServiceManagerMainDelegate::RunEmbedderProcess() {
            [141419]          chrome |                 content::ContentMainRunnerImpl::Run() {
            [141419]          chrome |                   content::RunOtherNamedProcessTypeMain() {
            [141419]          chrome |                     content::RunZygote() {
            [141419]          chrome |                       service_manager::ZygoteMain() {
            [141419]          chrome |                         service_manager::Zygote::ProcessRequests() {
            [141419]          chrome |                           ppoll() {
            [141419]          chrome |                             /* linux:sched-out */
   2.015  m [141297]          chrome |                       /* linux:sched-in */
   2.015  m [141297]          chrome |                     } /* read */
   2.015  m [141297]          chrome |                   } /* service_manager::ZygoteCommunication::GetSandboxStatus */
   2.015  m [141297]          chrome |                 } /* content::ContentMainRunnerImpl::Initialize */
   2.015  m [141297]          chrome |               } /* content::ContentServiceManagerMainDelegate::Initialize */
            [141297]          chrome |               content::ContentServiceManagerMainDelegate::RunEmbedderProcess() {
            [141297]          chrome |                 content::ContentMainRunnerImpl::Run() {
            [141297]          chrome |                   content::ContentMainRunnerImpl::RunServiceManager() {
            [141482]          chrome | base::_GLOBAL__N_1::ThreadFunc() {
            [141483]          chrome | base::_GLOBAL__N_1::ThreadFunc() {
            [141482]          chrome |   base::Thread::ThreadMain() {
            [141482] TaskSchedulerSe |     /* linux:task-name (name=TaskSchedulerSe) */
            [141483]          chrome |   base::internal::SchedulerWorker::ThreadMain() {
            [141483]          chrome |     base::internal::SchedulerWorker::RunPooledWorker() {
            [141483]          chrome |       base::internal::SchedulerWorker::RunWorker() {
            [141483] TaskSchedulerFo |         /* linux:task-name (name=TaskSchedulerFo) */
            [141483] TaskSchedulerFo |         base::internal::SchedulerWorker::Delegate::WaitForWork() {
            [141483] TaskSchedulerFo |           base::WaitableEvent::TimedWait() {
            [141483] TaskSchedulerFo |             base::WaitableEvent::TimedWaitUntil() {
            [141483] TaskSchedulerFo |               base::ConditionVariable::TimedWait() {
            [141483] TaskSchedulerFo |                 pthread_cond_timedwait() {
            [141483] TaskSchedulerFo |                   /* linux:sched-out */
...
            [141497]          chrome |                   base::debug::TaskAnnotator::RunTask() {
            [141497]          chrome |                     base::internal::Invoker::RunOnce() {
            [141497]          chrome |                       blink::ScriptRunner::ExecuteTask() {
            [141497]          chrome |                         blink::ScriptRunner::ExecuteAsyncTask() {
            [141497]          chrome |                           blink::PendingScript::ExecuteScriptBlock() {
            [141497]          chrome |                             blink::PendingScript::ExecuteScriptBlockInternal() {
            [141497]          chrome |                               blink::ClassicScript::RunScript() {
            [141497]          chrome |                                 blink::ScriptController::ExecuteScriptInMainWorld() {
            [141497]          chrome |                                   blink::ScriptController::EvaluateScriptInMainWorld() {
            [141497]          chrome |                                     blink::ScriptController::ExecuteScriptAndReturnValue() {
            [141497]          chrome |                                       blink::V8ScriptRunner::CompileScript() {
            [141497]          chrome |                                         blink::_GLOBAL__N_1::CompileScriptInternal() {
            [141497]          chrome |                                           v8::ScriptCompiler::Compile() {
            [141497]          chrome |                                             v8::internal::Compiler::GetSharedFunctionInfoForStreamedScript() {
   1.029  s [141499] TaskSchedulerFo |                   /* linux:sched-in */
   1.029  s [141499] TaskSchedulerFo |                 } /* pthread_cond_timedwait */
   1.029  s [141499] TaskSchedulerFo |               } /* base::ConditionVariable::TimedWait */
   1.029  s [141499] TaskSchedulerFo |             } /* base::WaitableEvent::TimedWaitUntil */
   1.029  s [141499] TaskSchedulerFo |           } /* base::WaitableEvent::TimedWait */
   1.029  s [141499] TaskSchedulerFo |         } /* base::internal::SchedulerWorker::Delegate::WaitForWork */
            [141499] TaskSchedulerFo |         base::internal::SchedulerWorker::Delegate::WaitForWork() {
            [141499] TaskSchedulerFo |           base::WaitableEvent::TimedWait() {
            [141499] TaskSchedulerFo |             base::WaitableEvent::TimedWaitUntil() {
            [141499] TaskSchedulerFo |               base::ConditionVariable::TimedWait() {
            [141499] TaskSchedulerFo |                 pthread_cond_timedwait() {
            [141499] TaskSchedulerFo |                   /* linux:sched-out */
...
            [141496]             exe | ChromeMain() {
            [141496]             exe |   headless::HeadlessShellMain() {
            [141496]             exe |     headless::HeadlessShellMain() {
            [141496]             exe |       headless::RunChildProcessIfNeeded() {
            [141496]          chrome |         /* linux:task-name (name=chrome) */
            [141496]          chrome |         exit() {
            [141496]          chrome |           /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 141297
[16] base::MessagePumpGlib::Run
[15] base::MessageLoopImpl::Run
[14] base::RunLoop::Run
[13] content::BrowserMainLoop::MainMessageLoopRun
[12] content::BrowserMainLoop::RunMainMessageLoopParts
[11] content::BrowserMainRunnerImpl::Run
[10] headless::HeadlessContentMainDelegate::RunProcess
[9] content::ContentMainRunnerImpl::RunServiceManager
[8] content::ContentMainRunnerImpl::Run
[7] content::ContentServiceManagerMainDelegate::RunEmbedderProcess
[6] service_manager::Main
[5] content::ContentMain
[4] headless::_GLOBAL__N_1::RunContentMain
[3] headless::HeadlessBrowserMain
[2] headless::HeadlessShellMain
[1] headless::HeadlessShellMain
[0] ChromeMain

task: 141418
[3] content::SandboxIPCHandler::Run
[2] base::DelegateSimpleThread::Run
[1] base::SimpleThread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141419
[12] service_manager::Zygote::ProcessRequests
[11] service_manager::ZygoteMain
[10] content::RunZygote
[9] content::RunOtherNamedProcessTypeMain
[8] content::ContentMainRunnerImpl::Run
[7] content::ContentServiceManagerMainDelegate::RunEmbedderProcess
[6] service_manager::Main
[5] content::ContentMain
[4] headless::_GLOBAL__N_1::RunContentMain
[3] headless::RunChildProcessIfNeeded
[2] headless::HeadlessShellMain
[1] headless::HeadlessShellMain
[0] ChromeMain

task: 141482
[5] base::MessagePumpLibevent::Run
[4] base::MessageLoopImpl::Run
[3] base::RunLoop::Run
[2] base::internal::ServiceThread::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141483
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunPooledWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141484
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunPooledWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141485
[36] base::debug::_GLOBAL__N_1::StackDumpSignalHandler
[35] bn_mul_mont
[34] BN_mod_mul_montgomery
[33] BN_mod_exp_mont
[32] RSA_verify_raw
[31] RSA_verify_pss_mgf1
[30] pkey_rsa_verify
[29] EVP_PKEY_verify
[28] EVP_DigestVerifyFinal
[27] EVP_DigestVerify
[26] bssl::ssl_public_key_verify
[25] bssl::ssl_client_handshake
[24] bssl::ssl_run_handshake
[23] SSL_do_handshake
[22] net::SSLClientSocketImpl::DoHandshake
[21] net::SSLClientSocketImpl::DoHandshakeLoop
[20] net::SSLClientSocketImpl::OnHandshakeIOComplete
[19] net::SSLClientSocketImpl::OnVerifyComplete
[18] base::internal::Invoker::RunOnce
[17] net::CachingCertVerifier::OnRequestFinished
[16] base::internal::Invoker::RunOnce
[15] net::CertVerifierJob::OnJobCompleted
[14] base::internal::Invoker::RunOnce
[13] base::internal::ReplyAdapter
[12] base::internal::Invoker::RunOnce
[11] base::internal::Invoker::RunOnce
[10] base::internal::Invoker::RunOnce
[9] base::debug::TaskAnnotator::RunTask
[8] base::MessageLoopImpl::RunTask
[7] base::MessageLoopImpl::DoWork
[6] base::MessagePumpLibevent::Run
[5] base::MessageLoopImpl::Run
[4] base::RunLoop::Run
[3] content::BrowserProcessSubThread::IOThreadRun
[2] content::BrowserProcessSubThread::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141486
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141488
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunPooledWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141489
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunPooledWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141491
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunDedicatedWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141493
[4] base::MessagePumpDefault::Run
[3] base::MessageLoopImpl::Run
[2] base::RunLoop::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141495
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunDedicatedWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141496
[4] exit
[3] headless::RunChildProcessIfNeeded
[2] headless::HeadlessShellMain
[1] headless::HeadlessShellMain
[0] ChromeMain

task: 141497
[3] base::MessagePumpDefault::Run
[2] base::MessageLoopImpl::Run
[1] base::RunLoop::Run
[0] content::RendererMain

task: 141499
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunPooledWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141500
[18] v8::internal::ParserBase::ParseBlock
[17] v8::internal::ParserBase::ParseStatement
[16] v8::internal::ParserBase::ParseStatementListItem
[15] v8::internal::Parser::DoParseProgram
[14] v8::internal::Parser::ParseOnBackground
[13] v8::internal::BackgroundCompileTask::Run
[12] v8::ScriptCompiler::ScriptStreamingTask::Run
[11] blink::_GLOBAL__N_1::RunScriptStreamingTask
[10] blink::_GLOBAL__N_1::RunBlockingScriptStreamingTask
[9] base::internal::Invoker::Run
[8] base::debug::TaskAnnotator::RunTask
[7] base::internal::TaskTracker::RunSkipOnShutdown
[6] base::internal::TaskTracker::RunOrSkipTask
[5] base::internal::TaskTrackerPosix::RunOrSkipTask
[4] base::internal::TaskTracker::RunAndPopNextTask
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunPooledWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141501
[9] mojo::core::NodeChannel::OnChannelError
[8] mojo::core::Channel::OnError
[7] mojo::core::_GLOBAL__N_1::ChannelPosix::OnFileCanReadWithoutBlocking
[6] base::MessagePumpLibevent::OnLibeventNotification
[5] event_base_loop
[4] base::MessagePumpLibevent::Run
[3] base::MessageLoopImpl::Run
[2] base::RunLoop::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141502
[4] base::MessagePumpDefault::Run
[3] base::MessageLoopImpl::Run
[2] base::RunLoop::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141503
[4] base::MessagePumpDefault::Run
[3] base::MessageLoopImpl::Run
[2] base::RunLoop::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141504
[3] base::internal::SchedulerWorker::RunWorker
[2] base::internal::SchedulerWorker::RunSharedWorker
[1] base::internal::SchedulerWorker::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

task: 141505
[4] base::MessagePumpDefault::Run
[3] base::MessageLoopImpl::Run
[2] base::RunLoop::Run
[1] base::Thread::ThreadMain
[0] base::_GLOBAL__N_1::ThreadFunc

```

**Comment by honggyukim** [comment]
As I mentioned before, we need to have `-Z` function size filter to better debug these kind of issues.

**Comment by honggyukim** [comment]
@ParkHanbum Could you please rebase [review/dynamic-trace-v6](https://github.com/namhyung/uftrace/compare/review/dynamic-trace-v6) and apply this change?

I think we better split `libmcount.so` and `libmcount-dynamic.so` to minimize overhead when dynamic feature is not used. So `libmcount-dynamic.so` has to restore all the registers while `libmcount.so` doesn't have to.

**Comment by honggyukim** [comment]
After testing more programs, I found that it’s quite stable enough although it has some problems for very exceptional cases. I think it’s almost ready to be merged into master.

**Comment by honggyukim** [comment]
@ParkHanbum Can you force push it into #588 again?

**Comment by namhyung** [comment]
Thank you very much for the analysis.  But I still don't understand why it makes the problem.  Anyway I'll update the dynamic tracing branch after more investigations.

**Comment by ParkHanbum** [comment]
> 
> 
> @ParkHanbum Could you please rebase [review/dynamic-trace-v6](https://github.com/namhyung/uftrace/compare/review/dynamic-trace-v6) and apply this change?
> 
sure, I'll update it. 

> I think we better split `libmcount.so` and `libmcount-dynamic.so` to minimize overhead when dynamic feature is not used. So `libmcount-dynamic.so` has to restore all the registers while `libmcount.so` doesn't have to.

To agree. I developed it separately at first. I was think it can merged between full-dynamic with previous libmcount. but it is mistake clearly.

**Comment by ParkHanbum** [comment]
> 
> 
> @ParkHanbum Can you force push it into #588 again?
> I would like to keep the discussion of [#588 (comment)](https://github.com/namhyung/uftrace/pull/588#issuecomment-445468114).
I think it is not good idea. something has been changed in review branch. we can discuss it continue in issue, aren't we?

**Comment by ParkHanbum** [comment]
> Thank you very much for the analysis. But I still don't understand why it makes the problem. Anyway I'll update the dynamic tracing branch after more investigations.

imagine sequence.
there is functions that share the variable allocated to registers between callee and caller.

A( arg *a) -> call B( arg *a) -> execute B()

after dynamic code patching, sequence has been change.

A( arg *a ) -> call B( arg *a ) -> mcount_entry() -> execute B()

Changes to variables shared by registers before and after the call to mcount_entry will affect future execution. So, why did not the problem happen before? Optimization to share variables with registers is not done automatically when profiling or the flag for instrument is set.

**Comment by ParkHanbum** [comment]
```
mcount: initializing mcount library
dwarf: prepare debug info
dynamic: dynamic patch type: 0 (none)
dynamic: dynamic update stats:
dynamic:    total:        1
dynamic:  patched:        1 (100.00%)
dynamic:   failed:        0 (0.00%)
dynamic:  skipped:        0 (0.00%)
dynamic: no match:        0
mcount: mcount setup done
[Switching to Thread 0x7ffff7fb6840 (LWP 2533)]

Thread 2.1 "uftrace" hit Breakpoint 1, __dentry__ () at /home/m/git/uftrace_origin/arch/x86_64/dynamic.S:45
45              sub $48, %rsp
mcount: new session started: 6f9e825450dbe297: uftrace

Thread 2.1 "uftrace" hit Breakpoint 2, __dentry__ () at /home/m/git/uftrace_origin/arch/x86_64/dynamic.S:104
104             retq
$1 = "restore register failed"
$2 = 0x0
$3 = 0x5555559e3bf0
$4 = 0x5555559e5b70
$5 = 0x5555559e5b70
$6 = 0x5555559cfd70
$7 = 0x5555559e5ae0
$8 = 0x0
$9 = 0x5555559e5b70
$10 = 0x5555559e5ae0
$11 = 0x5555559cfd70
$12 = 0x0
$13 = 0x7fffffffd080
$14 = 0x5555559e6000
$15 = 0x5555559e5b80
```

```
(gdb) info registers
rax            0x0      0
rbx            0x5555559e3bf0   93824997014512
rcx            0x5555559e5b70   93824997022576
rdx            0x5555559e5b70   93824997022576
rsi            0x5555559cfd70   93824996932976
rdi            0x5555559e5ae0   93824997022432
rbp            0x7fffffffd050   0x7fffffffd050
rsp            0x7fffffffce70   0x7fffffffce70
r8             0x0      0
r9             0x5555559e5b70   93824997022576
r10            0x7fffffffcd80   140737488342400
r11            0x1      1
r12            0x0      0
r13            0x7fffffffd080   140737488343168
r14            0x5555559e6000   93824997023744
r15            0x5555559e5b80   93824997022592
rip            0x7ffff7bbfc18   0x7ffff7bbfc18 <__dentry__+117>
eflags         0x202    [ IF ]
cs             0x33     51
ss             0x2b     43
ds             0x0      0
es             0x0      0
fs             0x0      0
gs             0x0      0
```

as you can see some registers are not restored.

```
(gdb) c
Continuing.

Thread 2.1 "uftrace" received signal SIGSEGV, Segmentation fault.
rb_insert_color (node=0x5555559e3bf0, root=0x1, root@entry=0x5555559cfd70) at /home/m/git/uftrace_origin/utils/rbtree.c:133
133             rb_set_black(root->rb_node);
```
It causes segmentation-fault.


```
(gdb) x/i $rip
=> 0x55555558daaa <rb_insert_color+106>:        mov    (%r11),%rax
```

because value of register r11 has been change to 0x1 in __dentry__.
you can check it with following gdb script.

```
$ cat cmd.gdb
file uftrace
set follow-fork-mode child
set breakpoint pending on

set $save_reg_rax = 0
set $save_reg_rbx = 0
set $save_reg_rcx = 0
set $save_reg_rdx = 0
set $save_reg_rsi = 0
set $save_reg_rdi = 0
set $save_reg_r8 = 0
set $save_reg_r9 = 0
set $save_reg_r10 = 0
set $save_reg_r11 = 0
set $save_reg_r12 = 0
set $save_reg_r13 = 0
set $save_reg_r14 = 0
set $save_reg_r15 = 0

br arch/x86_64/dynamic.S:45
commands
        set $save_reg_rax = $rax
        set $save_reg_rbx = $rbx
        set $save_reg_rcx = $rcx
        set $save_reg_rdx = $rdx
        set $save_reg_rsi = $rsi
        set $save_reg_rdi = $rdi
        set $save_reg_r8 = $r8
        set $save_reg_r9 = $r9
        set $save_reg_r10 = $r10
        set $save_reg_r11 = $r11
        set $save_reg_r12 = $r12
        set $save_reg_r13 = $r13
        set $save_reg_r14 = $r14
        set $save_reg_r15 = $r15
        continue
end

br arch/x86_64/dynamic.S:104
commands
  if $save_reg_rax == $rax & $save_reg_rbx == $rbx & $save_reg_rcx == $rcx & $save_reg_rdx == $rdx & $save_reg_rsi == $rsi & $save_reg_rdi == $rdi & $save_reg_r8 == $r8 & $save_reg_r9 == $r9 & $save_reg_r10 == $
r10 & $save_reg_r11 == $r11 & $save_reg_r12 == $r12 & $save_reg_r13 == $r13 & $save_reg_r14 == $r14 & $save_reg_r15 == $r15
    continue
  else
    p "restore register failed"
    print_reg
  end
end

r record -L. -P__rb_rotate_right -v -la --no-libcall ./uftrace replay -d abc --no-pager
```

If you want to know where you changed the r11 register, you can add watch as you know.

```
br arch/x86_64/dynamic.S:45
commands
        set $save_reg_rax = $rax
        set $save_reg_rbx = $rbx
        set $save_reg_rcx = $rcx
        set $save_reg_rdx = $rdx
        set $save_reg_rsi = $rsi
        set $save_reg_rdi = $rdi
        set $save_reg_r8 = $r8
        set $save_reg_r9 = $r9
        set $save_reg_r10 = $r10
        set $save_reg_r11 = $r11
        set $save_reg_r12 = $r12
        set $save_reg_r13 = $r13
        set $save_reg_r14 = $r14
        set $save_reg_r15 = $r15
# here 
        **watch $r11 == 0x1**
        continue
end
```

**Comment by honggyukim** [comment]
Hmm.. I cannot reproduce your gdb script and it works fine in my environment.

**Comment by ParkHanbum** [comment]
@honggyukim

**Comment by ParkHanbum** [comment]
I have wrote a perfect example for this issue.
and I think it will makes more understandable than my explain.

example : 
```
/* { dg-do run } */
/* { dg-options "-fipa-ra" } */
/* Testing -fipa-ra optimization option.  */
#include <assert.h>

static int __attribute__((noinline))
bar (int a, int b, int c, int d, int e, int f)
{
  asm("nopl   (%ebx,%ebx,1)");
  asm("nopl   (%ebx,%ebx,1)");
  return a + b + c + d + e + f + 3;
}

int __attribute__((noinline))
foo (int a, int b, int c, int d, int e, int f)
{
  asm("nopl   (%ebx,%ebx,1)");
  asm("nopl   (%ebx,%ebx,1)");

  int a_dup = a;
  int b_dup = b;
  a_dup += bar (a,b_dup,c,d,e,f);

  return a_dup + b_dup;
}
```
int
main (void)
{
  asm("nopl   (%ebx,%ebx,1)");
  asm("nopl   (%ebx,%ebx,1)");

  assert(foo(1,1,1,1,1,1) == 11);
}
```

compile this like this: 
```
$ gcc -g -O2 -fipa-ra fuse-caller-save.c
```

this is results :
```
# exit normally
$ ./a.out 

$ uftrace record -P. -v a.out                                                                              
[13/42]uftrace: removing uftrace.data.old directory
uftrace: skipping perf event due to error: Permission denied
uftrace: please check /proc/sys/kernel/perf_event_paranoid
uftrace: creating 1 thread(s) for recording
uftrace: using libmcount.so library for tracing
mcount: initializing mcount library
dwarf: prepare debug info
dynamic: dynamic patch type: 0 (none)
dynamic: dynamic update stats:
dynamic:    total:        3
dynamic:  patched:        3 (100.00%)
dynamic:   failed:        0 (0.00%)
dynamic:  skipped:        0 (0.00%)
dynamic: no match:        0
plthook: setup PLT hooking "/home/m/git/a.out"
mcount: mcount setup done
mcount: new session started: f91f5bdb4823e465: a.out
a.out: fuse-caller-save.c:33: main: Assertion `foo(1,1,1,1,1,1) == 11' failed.
process crashed by signal 6: Aborted (si_code: -6)
Backtrace from uftrace:
=====================================
[1] (__assert_fail[55b2c9728520] <= <55b2c97285ca>[55b2c97285ca])
[0] (main[55b2c9728545] <= __libc_start_main[7f114c193b97])
child terminated by signal: 6: Aborted
uftrace: reading uftrace.data/task.txt file
uftrace: flushing /uftrace-f91f5bdb4823e465-7678-000
```

**Comment by honggyukim** [comment]
@ParkHanbum I can also reproduce the problem and it can be fixed by restoring all the registers as you posted above. I will take a look when I have some time.

It's really helpful to write a small reproducible example code. Thanks!

**Comment by ParkHanbum** [comment]
> 
> 
> @ParkHanbum I can also reproduce the problem and it can be fixed by restoring all the registers as you posted above. I will take a look when I have some time.
> 
> It's really helpful to write a small reproducible example code. Thanks!

how about cgdb?

**Comment by honggyukim** [comment]
> how about cgdb?

What do you mean?

**Comment by ParkHanbum** [comment]
> 
> 
> @honggyukim I don't know what how cgdb work. but I have some theory.
> 
> ```
> $ uftrace record -P. `which cgdb` -ex="b main" -ex="r" -ex="q" hello
> ```
> 
> move breakpoint to middle of function like 'main+10'.
> if my theory is right, it works.
> 
> My gdb script is too limited to my environment. I will modify it for use in other common environments.

@honggyukim

**Comment by namhyung** [comment]
@ParkHanbum thanks for the reproducer.  It seems the problem comes from `r10` and `r11` registers which are **NOT** preserved across function calls.  But with `-fipa-ra` compiler seems to generate code that should be preserved.  I guess other registers are ok since they are said to be preserved already.

**Comment by ParkHanbum** [comment]
@namhyung yes, but It is not limited to r10 and r11. Check out the tests I added.
https://github.com/namhyung/uftrace/pull/733/commits/5ac031bfc1d945138345394bf8f4a767b8d71280

**Comment by namhyung** [comment]
I saw the test but why do you think it's not?

**Comment by ParkHanbum** [comment]
all of regsters are allocated as variable in function foo.

**Comment by namhyung** [comment]
I don't think like that.

**Comment by namhyung** [comment]
And other registers will be preserved after function call.

**Comment by namhyung** [comment]
> And other registers will be preserved after function call.

Oh I meant callee-saved registers like `rbp`, `rbx`, `r12`, ...  So I think it's ok to add just` r10` and `r11` for entry, but exit pass should save and restore other **caller**-saved registers like `rsi`, `rdi`, ... like entry does.

**Comment by namhyung** [comment]
@honggyukim could you please test [below patch](https://github.com/namhyung/uftrace/files/3119427/save-more-registers.patch.txt)?

```diff
diff --git a/arch/x86_64/dynamic.S b/arch/x86_64/dynamic.S
index ec19260..a384968 100644
--- a/arch/x86_64/dynamic.S
+++ b/arch/x86_64/dynamic.S
@@ -69,10 +69,14 @@ GLOBAL(__dentry__)
 	/* save rax (implicit argument for variadic functions) */
 	push %rax
 
+	/* save scratch registers due to -fipa-ra */
+	push %r10
+	push %r11
+
 	call mcount_entry
 
 	/* original stack pointer */
-	movq 8(%rsp), %rdx
+	movq 24(%rsp), %rdx
 
 	/* child addr */
 	movq 48(%rdx), %rdi
@@ -81,11 +85,13 @@ GLOBAL(__dentry__)
 	call mcount_find_code
 
 	/* original stack pointer */
-	movq 8(%rsp), %rdx
+	movq 24(%rsp), %rdx
 
 	/* overwrite return address */
 	movq %rax, 48(%rdx)
 
+	pop  %r11
+	pop  %r10
 	pop  %rax
 
 	movq %rdx, %rsp
diff --git a/arch/x86_64/mcount.S b/arch/x86_64/mcount.S
index 283b9c2..a12cf0a 100644
--- a/arch/x86_64/mcount.S
+++ b/arch/x86_64/mcount.S
@@ -84,10 +84,16 @@ END(mcount)
 
 ENTRY(mcount_return)
 	.cfi_startproc
-	sub $48, %rsp
-	.cfi_def_cfa_offset 48
-
-	movdqu %xmm0, 16(%rsp)
+	sub $80, %rsp
+	.cfi_def_cfa_offset 80
+
+	movq   %r11,  64(%rsp)
+	movq   %r10,  56(%rsp)
+	movq   %r9,   48(%rsp)
+	movq   %r8,   40(%rsp)
+	movq   %rdi,  32(%rsp)
+	movq   %rsi,  24(%rsp)
+	movq   %rcx , 16(%rsp)
 	movq   %rdx,   8(%rsp)
 	movq   %rax,   0(%rsp)
 
@@ -96,13 +102,19 @@ ENTRY(mcount_return)
 
 	/* returns original parent address */
 	call mcount_exit
-	movq %rax, 40(%rsp)
+	movq %rax, 72(%rsp)
 
 	movq    0(%rsp), %rax
 	movq    8(%rsp), %rdx
-	movdqu 16(%rsp), %xmm0
-
-	add $40, %rsp
+	movq   16(%rsp), %rcx
+	movq   24(%rsp), %rsi
+	movq   32(%rsp), %rdi
+	movq   40(%rsp), %r8
+	movq   48(%rsp), %r9
+	movq   56(%rsp), %r10
+	movq   64(%rsp), %r11
+
+	add $72, %rsp
 	.cfi_def_cfa_offset 8
 	retq
 	.cfi_endproc
```

**Comment by honggyukim** [comment]
Sure. Here is the tested programs. As a result, everything works fine except for chrome same as before.
1. uftrace
```
$ uftrace record -P. -a ./uftrace replay -d abc
uftrace: some functions cannot be patched dynamically (619/749)
uftrace: some functions cannot be patched dynamically (0/0)
# DURATION     TID     FUNCTION
   1.440 us [ 72663] | __monstartup();
   0.913 us [ 72663] | __cxa_atexit();
            [ 72663] | main() {
            [ 72663] |   a() {
            [ 72663] |     b() {
            [ 72663] |       c() {
   0.717 us [ 72663] |         getpid();
   1.703 us [ 72663] |       } /* c */
   2.184 us [ 72663] |     } /* b */
   2.570 us [ 72663] |   } /* a */
   3.023 us [ 72663] | } /* main */
```
2. node
```
$ uftrace record -P. -a ./node --expose-gc --trace-gc -e 'gc()'
uftrace: some functions cannot be patched dynamically (35276/37886)
[98626:0xc492e50]      820 ms: Scavenge 2.4 (4.2) -> 2.1 (5.2) MB, 14.4 / 0.1 ms  (average mu = 1.000, current mu = 1.000) allocation failure
[98626:0xc492e50]     1114 ms: Scavenge 2.6 (5.2) -> 2.5 (6.2) MB, 14.2 / 0.0 ms  (average mu = 1.000, current mu = 1.000) allocation failure
[98626:0xc492e50]     1671 ms: Mark-sweep 3.5 (6.2) -> 3.2 (8.2) MB, 30.3 / 0.0 ms  (average mu = 1.000, current mu = 1.000) testing GC in old space requested
```
3. clang
```
$ uftrace record -P. -a ./clang ~/hello.c
uftrace: some functions cannot be patched dynamically (44453/53174)
uftrace: some functions cannot be patched dynamically (44453/53174)
```
4. chrome
```
$ uftrace record -P. -t 1ms ./chrome --no-sandbox --headless https://www.google.com
uftrace: some functions cannot be patched dynamically (285354/326211)
uftrace: some functions cannot be patched dynamically (285354/326211)
Received signal 4 ILL_ILLOPN 55a23eed6164
#0 0x7fcfa5c08e47 backtrace
#1 0x55a2411c614f (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca214e)
#2 0x55a2411c5cb1 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca1cb0)
#3 0x7fcfa57ee390 <unknown>
#4 0x55a23eed6164 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x29b2163)
#5 0x55a241a60bde (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x553cbdd)
#6 0x55a241a6040d (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x553c40c)
#7 0x55a241a8e7db (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x556a7da)
#8 0x55a241a8f3b9 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x556b3b8)
#9 0x55a241a57c01 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5533c00)
#10 0x55a241a54936 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5530935)
#11 0x55a241a54a81 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5530a80)
#12 0x55a241ae8753 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55c4752)
#13 0x55a241ae4791 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55c0790)
#14 0x55a241adc0fb (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55b80fa)
#15 0x55a241ad9804 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x55b5803)
#16 0x55a241842ff7 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x531eff6)
#17 0x55a241841d3c (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x531dd3b)
#18 0x55a2418441c7 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x53201c6)
#19 0x55a241864701 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5340700)
#20 0x55a241864b9c (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5340b9b)
#21 0x55a241877732 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5353731)
#22 0x55a241877cb1 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x5353cb0)
#23 0x55a2418778b2 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x53538b1)
#24 0x55a23eef9527 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x29d5526)
#25 0x55a23f8404fe (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x331c4fd)
#26 0x55a2411b4f5a (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c90f59)
#27 0x55a2411197c3 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4bf57c2)
#28 0x55a24113591f (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c1191e)
#29 0x55a241135fe2 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c11fe1)
#30 0x55a2411d80e9 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4cb40e8)
#31 0x55a241156539 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c32538)
#32 0x55a23f76ac34 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x3246c33)
#33 0x55a24119a1e2 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4c761e1)
#34 0x55a2411db46e (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4cb746d)
#35 0x7fcfa57e46ba start_thread
#36 0x7fcf9f75b41d clone
  r8: 0000132ca817bf38  r9: 0000000000000020 r10: 0000000000000000 r11: 0000000000080100
 r12: 0000000000000020 r13: 0000132ca817bf38 r14: 0000132cea2c2768 r15: 0000132ca817bf08
  di: 0000132cea26eea8  si: 0000132cea26e248  bp: 00007fceaa9d0a50  bx: 0000132cea2c2708
  dx: 0000132cea04b248  ax: 00007fceaa9d0a18  cx: 0000132cea26ed88  sp: 00007fceaa9d0a10
  ip: 000055a23eed6164 efl: 0000000000010246 cgf: 0000000000000033 erf: 0000000000000000
 trp: 0000000000000006 msk: 0000000000000000 cr2: 0000000000000000
[end of stack trace]
Calling _exit(1). Core file will not be generated.
uftrace: some functions cannot be patched dynamically (285354/326211)
```

**Comment by honggyukim** [comment]
It's suspicious that the broken log shows it calls `backtrace` function right after `ILL_ILLOPN` shown.
```
$ uftrace record -P. -t 1ms ./chrome --no-sandbox --headless https://www.google.com
uftrace: some functions cannot be patched dynamically (285354/326211)
uftrace: some functions cannot be patched dynamically (285354/326211)
Received signal 4 ILL_ILLOPN 55a23eed6164
#0 0x7fcfa5c08e47 backtrace
#1 0x55a2411c614f (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca214e)
#2 0x55a2411c5cb1 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x4ca1cb0)
#3 0x7fcfa57ee390 <unknown>
#4 0x55a23eed6164 (/home/honggyu/work/chromium/master/src/out/Release/chrome+0x29b2163)
    ...
```

**Comment by ParkHanbum** [comment]
@namhyung I think it is not good idea. how we can assume that only r10 and r11 be changed inner mcount_entry and mcount_exit? in this case, r11 register was changed in malloc at libc. There is always the possibility that variables can allocate to register except r10 and r11 as like my committed test.

**Comment by namhyung** [comment]
Since we already saves most (caller-saved) registers in `mcount_entry()` and a few in `mcount_return()`.
Other (callee-saved) registers will not be changed according to the **calling convention**.

In my understanding, this is a problem about (caller-saved) scratch registers.  By the calling convention, values in the scratch registers can (and likely to) be changed after function call.  But with `-fipa-ra` compiler can know which register is changed or not, and it could use old value without saving and restoring it explicitly after function call.

**Comment by namhyung** [comment]
@honggyukim thanks for the test! For chrome, it might be a different problem like binary pattern or something.  I'll update the patch series with this result.

**Comment by honggyukim** [comment]
I also think that it looks more like a corner case and can be investigated later. I'm okay with the current status. Thanks.

**Comment by ParkHanbum** [comment]
So let's take an example.
- The target program uses the r12 register as you said.
(Because the call to the dentry function can not be computed at compile time).
- The dentry function call does not restore the r12 register.
(As if it had not been previously restored with r10 and r11 registers)
- The same problem occurs when the r12 register changes in dentry.

And we can not guarantee that the r12 register will not change. Previously I thought the r10 and r11 registers would not change, but it's not.

**Comment by ParkHanbum** [comment]
If not changed the register according to the calling convention, you should not change r10 and r11 as well. Is not it?

**Comment by namhyung** [comment]
@ParkHanbum please read the calling convention - r11 and r12 are different.  gcc will do the job for us for the callee-saved registers (if needed).

And I didn't think the r11 is not changed.  I thought code won't access r11 after function call, but it is (with `-fipa-ra`).  The same goes to the other caller-saved registers (like rsi, ...).

**Comment by ParkHanbum** [comment]
you misunderstood. According to the function calling convention, there is no obligation to restore functions that are called other than six registers. So malloc called by mcount_entry changed the r11 register and did not restore it.

**Comment by ParkHanbum** [comment]
And for that reason, if it force the -ffixed option to not use a particular register, it do not get an error.

**Comment by ParkHanbum** [comment]
Currently, restoring r12 to r15 from mcount_entry is done because allocating the corresponding registers inside mcount_entry. This is not guaranteed.

**Comment by honggyukim** [comment]
>  I don't know what how cgdb work. but I have some theory.
> ```
> $ uftrace record -P. `which cgdb` -ex="b main" -ex="r" -ex="q" hello
> ```
> move breakpoint to middle of function like 'main+10'.
> if my theory is right, it works.

I tried again with `cgdb` and `gdb`, but the problem is still there even if I set the breakpoint at `main+10` as you suggested.
```asm
0000000000400526 <main>:
  400526:       55                      push   %rbp
  400527:       48 89 e5                mov    %rsp,%rbp
  40052a:       bf c4 05 40 00          mov    $0x4005c4,%edi    # here is the breakpoint address for main
  40052f:       e8 cc fe ff ff          callq  400400 <puts@plt>
  400534:       b8 00 00 00 00          mov    $0x0,%eax         # breakpoint of "main+10"
  400539:       5d                      pop    %rbp
  40053a:       c3                      retq
  40053b:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
```
```
$ uftrace record -P. `which gdb` -q -ex="b *0x400534" -ex="r" -ex="q" hello
uftrace: some functions cannot be patched dynamically (4013/5527)
Reading symbols from hello...(no debugging symbols found)...done.
Breakpoint 1 at 0x400534
Starting program: /home/honggyu/work/uftrace/hello 
uftrace: some functions cannot be patched dynamically (766/1463)
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Hello World!

Breakpoint 1, 0x0000000000400534 in main ()
A debugging session is active.

        Inferior 1 [process 103329] will be killed.

Quit anyway? (y or n) y
(... running forever ...)
```

**Comment by ParkHanbum** [comment]
@honggyukim I thought that dynamic patch might overwritten the CR3. I'm wrong. thank you for testing.

**Comment by honggyukim** [comment]
@ParkHanbum Thanks for your analysis as well.

**Comment by namhyung** [comment]
> Currently, restoring r12 to r15 from mcount_entry is done because allocating the corresponding registers inside mcount_entry. This is not guaranteed.

Why do you think it's not guaranteed?  That's what callee-saved register is for and compiler will generate code to save and restore if it's used.

**Comment by ParkHanbum** [comment]
your word is exactly what I want to say. If there is not need to restore the registers, then it will not be restored. is there any guarantee that r12 ~ r15 will always be used inside mcount_entry?

**Comment by namhyung** [comment]
If it's not used, it's unchanged.  So no need to restore, right?

**Comment by ParkHanbum** [comment]
In the previous error case, the r11 register is not used and is not restored, but its value has changed inside mcount_entry. did you mean that is exceptional case?

**Comment by namhyung** [comment]
https://stackoverflow.com/questions/18024672/what-registers-are-preserved-through-a-linux-x86-64-function-call

**Comment by ParkHanbum** [comment]
@namhyung I'm sorry about my misunderstand. I could not image that compiler use just only r10 and r11 register for specific purpose. it is my epic fail.

**Comment by honggyukim** [comment]
I found a few more error cases, which are not related to this problem.
Let's make further discussion in #281.

**Comment by namhyung** [comment]
@ParkHanbum It's not about they are used for specific purpose and in this case they're not.  It's not about which registers are protected by us (caller) or compiler (callee).
